clmock/clmock.go

```diff
@@ -0,0 +1,222 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package clmock
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/eth/catalyst"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type CLMock struct {
+	ctx               context.Context
+	cancel            context.CancelFunc
+	stack             *node.Node
+	eth               *eth.Ethereum
+	blockPeriod       time.Duration
+	withdrawals       []*types.Withdrawal
+	feeRecipient      common.Address
+	feeRecipientMutex sync.Mutex
+	nextWithdrawalIdx uint64
```

Comment:
Seems like you could abstract this a bit into another type `withdrawalGenerator` and inside you could have this var as `next`.

Reply:
Right now, I just don't see how abstracting this into a separate type changes much here (other than adding some boilerplate).  All that's happening is that a counter is being incremented every time a new withdrawal is received, and that this counter is used to validate incoming withdrawals.

Now if validation of withdrawals in this code changes and becomes more complicated, I could see value in separating that logic into a separate type.

Reply:
IMO it's more about naming convertions. `nextWithdrawalIdx`, `popWithdrawals` etc are pretty long, so I think a way to simplify a lot of the code here would be to make it it's own object. I was thinking something more along these lines: https://play.golang.com/p/Rbt6XQ8gpv_H